diff --git a/kernel/file.h b/kernel/file.h
index aa6d208..cddfe7e 100644
--- a/kernel/file.h
+++ b/kernel/file.h
@@ -2,7 +2,7 @@
 #define FILE_H
 
 struct file {
-  enum { FD_NONE, FD_PIPE, FD_INODE, FD_DEVICE } type;
+  enum { FD_NONE, FD_PIPE, FD_INODE, FD_DEVICE, FD_MUTEX} type;
   int ref; // reference count
   char readable;
   char writable;
@@ -11,6 +11,10 @@ struct file {
   uint off;          // FD_INODE and FD_DEVICE
   short major;       // FD_DEVICE
   short minor;       // FD_DEVICE
+
+  // ---------------------
+  struct sleeplock mutex;
+  // ---------------------
 };
 
 #define major(dev)  ((dev) >> 16 & 0xFFFF)
